### PR TITLE
Make changing default excludes at runtime an error

### LINKED ISSUE #15543: Remove deprecated PatternSpecFactory.reportChangedDefaultExcludes()
### Summary

- [x] `PatternSpecFactory.reportChangedDefaultExcludes()`

### Checklist
- [x] Validate whether we should remove the API in the 7.0 release
  - If the removal is not possible then create a follow-up issue and link it in the [overview spreadsheet](https://docs.google.com/spreadsheets/d/19J1nR_dFKpfKdu5KDFMVZGfjR0ysT9DthsBUPwf8mkM/edit#gid=493905624) 
- [x] Update the upgrade guide
- [x] Check User Manual for dangling references and outdated documentation
- [ ] Mark the story as done in the [overview spreadsheet](https://docs.google.com/spreadsheets/d/19J1nR_dFKpfKdu5KDFMVZGfjR0ysT9DthsBUPwf8mkM/edit#gid=2040581133).

### PR DESCRIPTION
**Pull request by melix**
This commit changes the deprecated behavior of changing the default
excludes at runtime, to make it an error instead.

Closes #15543

### CODE REVIEW

**Comment by melix** [comment]
@bot-gradle test this

**Comment by melix** [comment]
@bot-gradle test this

File: subprojects/core-api/src/main/java/org/gradle/api/tasks/util/internal/PatternSpecFactory.java [hunk 1]
```diff
@@ -85,26 +85,25 @@
...
         }
 
         return defaultExcludeSpecCache.get(caseSensitivity);
     }
 
+    protected boolean invalidChangeOfExcludes(String[] defaultExcludes) {
+        return !Arrays.equals(previousDefaultExcludes, defaultExcludes);
+    }
+
     private void reportChangedDefaultExcludes(String[] excludesFromSettings, String[] newDefaultExcludes) {
```
Comment (wolfs) [inline subprojects/core-api/src/main/java/org/gradle/api/tasks/util/internal/PatternSpecFactory.java:101]: This method should be called something like `failOnChangedDefaultExcludes()`.

File: subprojects/core-api/src/test/groovy/org/gradle/api/tasks/util/PatternSetTest.groovy [hunk 2]
```diff
@@ -16,18 +16,20 @@
...
+import org.gradle.api.tasks.util.internal.PatternSpecFactory
 import spock.lang.Issue
 
 import static org.gradle.util.Matchers.strictlyEquals
 import static org.hamcrest.CoreMatchers.equalTo
 import static org.hamcrest.MatcherAssert.assertThat
 
 class PatternSetTest extends AbstractTestForPatternSet {
-    PatternSet patternSet = new PatternSet()
+    PatternSet patternSet = new TestPatternSet()
```
Comment (wolfs) [inline subprojects/core-api/src/test/groovy/org/gradle/api/tasks/util/PatternSetTest.groovy:32]: I think this test should be fixed differently:
- Every time the test updates `DirectoryScanner.defaultExcludes`, it also needs to call `PatternSpecFactory.INSTANCE.setDefaultExcludesFromSettings(DirectoryScanner.getDefaultExcludes())`
- The cleanup also need to call `setDefaultExcludesFromSettings()` after resetting the default excludes
- There needs to be a new unit test which shows that using the spec fails when the default excludes changed without calling `setDefaultExcludesFromSettings()`.

**Review by wolfs — COMMENTED** [review]
I think we should adjust the PatternSpecFactory test so that it reflects what happens in the production code.

**Comment by melix** [comment]
@bot-gradle test QFL

File: subprojects/core-api/src/main/java/org/gradle/api/tasks/util/internal/PatternSpecFactory.java [hunk 3]
```diff
@@ -113,7 +114,12 @@ public synchronized void setDefaultExcludesFromSettings(String[] excludesFromSet
         }
     }
 
-    private void updateDefaultExcludeSpecCache(String[] defaultExcludes) {
+    @VisibleForTesting
+    public synchronized List<String> getDefaultExcludesFromSettings() {
```
Comment (wolfs) [inline subprojects/core-api/src/main/java/org/gradle/api/tasks/util/internal/PatternSpecFactory.java:118]: I don't think you need to expose this method. You can call `DirectoryScanner.getDefaultExcludes()` from the test.

File: subprojects/core-api/src/main/java/org/gradle/api/tasks/util/internal/PatternSpecFactory.java [hunk 4]
```diff
@@ -113,7 +114,12 @@ public synchronized void setDefaultExcludesFromSettings(String[] excludesFromSet
...
+        return ImmutableList.copyOf(previousDefaultExcludes);
+    }
+
+    protected void updateDefaultExcludeSpecCache(String[] defaultExcludes) {
```
Comment (wolfs) [inline subprojects/core-api/src/main/java/org/gradle/api/tasks/util/internal/PatternSpecFactory.java:122]: This method can remain `private`, right?

File: subprojects/core-api/src/main/java/org/gradle/api/tasks/util/internal/PatternSpecFactory.java [hunk 5]
```diff
@@ -85,26 +87,25 @@
         String[] defaultExcludes = DirectoryScanner.getDefaultExcludes();
         if (defaultExcludeSpecCache.isEmpty()) {
```
Comment (wolfs) [inline subprojects/core-api/src/main/java/org/gradle/api/tasks/util/internal/PatternSpecFactory.java:88]: Can we simplify this logic now? So if the cache is empty we populate it, and if it isn't, we verify that the default excludes didn't change. We don't need to update the spec cache the default excludes didn't change and the cache is non-empty, right?

File: subprojects/core-api/src/main/java/org/gradle/api/tasks/util/internal/PatternSpecFactory.java (hunk 4 above)
Comment (melix) [reply subprojects/core-api/src/main/java/org/gradle/api/tasks/util/internal/PatternSpecFactory.java:122]: Yes we can!

File: subprojects/core-api/src/main/java/org/gradle/api/tasks/util/internal/PatternSpecFactory.java (hunk 3 above)
Comment (melix) [reply subprojects/core-api/src/main/java/org/gradle/api/tasks/util/internal/PatternSpecFactory.java:118]: `DirectoryScanner.getDefaultExcludes()` would return the empty list in this case, where I need the "previous value".

**Comment by melix** [comment]
@bot-gradle test QFL

File: subprojects/core-api/src/main/java/org/gradle/api/tasks/util/internal/PatternSpecFactory.java (hunk 3 above)
Comment (wolfs) [reply subprojects/core-api/src/main/java/org/gradle/api/tasks/util/internal/PatternSpecFactory.java:116]: You can call `DirectoryScanner.reset...` to return to the default value.

Comment (melix) [reply subprojects/core-api/src/main/java/org/gradle/api/tasks/util/internal/PatternSpecFactory.java:116]: I'm not sure what you mean: this method is used in a test to return the list of default excludes configured in settings. This is only used to print the list. I don't use them to reset :thinking:

Comment (wolfs) [reply subprojects/core-api/src/main/java/org/gradle/api/tasks/util/internal/PatternSpecFactory.java:116]: In your test, you don't need this method. You can remember the previous state from `DirectoryScanner.getDefaultExcludes()` to check the exception message. I would prefer removing this method from `PatternSpecFactory`.

Comment (melix) [reply subprojects/core-api/src/main/java/org/gradle/api/tasks/util/internal/PatternSpecFactory.java:116]: ah now I see what you mean :)

**Comment by melix** [comment]
@bot-gradle test QFL

File: subprojects/snapshots/src/integTest/groovy/org/gradle/internal/vfs/DefaultExcludesIntegrationTest.groovy [hunk 6]
```diff
@@ -136,15 +136,10 @@ class DefaultExcludesIntegrationTest extends AbstractIntegrationSpec{
...
-            "Default excludes changed from ${defaultExcludesFromSettings} to ${defaultExcludesInTask}. " +
-            "Configure default excludes in the settings script instead. " +
-            "Consult the upgrading guide for further information: https://docs.gradle.org/current/userguide/upgrading_version_6.html#changing_default_excludes_during_the_execution_phase")
-        run "copyTask"
+        fails "copyTask"
+
         then:
-        executedAndNotSkipped(":copyTask")
-        copyOfExcludedFile.exists()
+        failure.assertHasCause "Cannot change default excludes during the build. They were changed from ${defaultExcludesFromSettings} to ${defaultExcludesInTask}. Configure default excludes in the settings script instead."
```
Comment (wolfs) [inline subprojects/snapshots/src/integTest/groovy/org/gradle/internal/vfs/DefaultExcludesIntegrationTest.groovy:142]: I think we also need to add something to the upgrading guide, right?

**Review by wolfs — COMMENTED** [review]
Looks good now. I think we miss some documentation.

File: subprojects/snapshots/src/integTest/groovy/org/gradle/internal/vfs/DefaultExcludesIntegrationTest.groovy (hunk 6 above)
Comment (melix) [reply subprojects/snapshots/src/integTest/groovy/org/gradle/internal/vfs/DefaultExcludesIntegrationTest.groovy:142]: actually it's already in there, but I missed that I have to move it from deprecations to migration.

**Review by wolfs — APPROVED** [review]
LGTM!

**Pull request merged**